nexus/db-queries/src/db/datastore/ipv4_nat_entry.rs

```diff
@@ -0,0 +1,495 @@
+        // constraint (which is desired in this case).
+        //
+        // It seems that this isn't straightforward to express in diesel
+        sql_query(
+            "
+            INSERT INTO omicron.public.ipv4_nat_entry (
+                external_address,
+                first_port,
+                last_port,
+                sled_address,
+                vni,
+                mac
+            )
+            SELECT external_address, first_port, last_port, sled_address, vni, mac
+            FROM ( VALUES ($1, $2, $3, $4, $5, $6) ) AS data (external_address, first_port, last_port, sled_address, vni, mac)
+            WHERE NOT EXISTS (
+                SELECT external_address, first_port, last_port, sled_address, vni, mac
+                FROM omicron.public.ipv4_nat_entry as entry
+                WHERE data.external_address = entry.external_address
+                AND data.first_port = entry.first_port
+                AND data.last_port = entry.last_port
+                AND data.sled_address = entry.sled_address
+                AND data.vni = entry.vni
+                AND data.mac = entry.mac
+            )
+            ",
+        )
+            .bind::<Inet, _>(nat_entry.external_address)
+            .bind::<Integer, _>(nat_entry.first_port)
+            .bind::<Integer, _>(nat_entry.last_port)
+            .bind::<Inet, _>(nat_entry.sled_address)
+            .bind::<Integer, _>(nat_entry.vni)
+            .bind::<BigInt, _>(nat_entry.mac)
+            .execute_async(&*self.pool_connection_authorized(opctx).await?)
+            .await
+            .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))?;
+
+        Ok(())
+    }
+
+    pub async fn ipv4_nat_delete(
+        &self,
+        opctx: &OpContext,
+        nat_entry: &Ipv4NatEntry,
+    ) -> DeleteResult {
+        // We use pure SQL here so we can call nextval() to increment the sequence number in the
+        // database.
+        let updated_rows = sql_query(
+            "
+            UPDATE omicron.public.ipv4_nat_entry
+            SET
+            version_removed = nextval('omicron.public.ipv4_nat_version'),
+            time_deleted = now()
+            WHERE time_deleted IS NULL AND version_removed IS NULL AND id = $1 AND version_added = $2
+            ",
+        )
+            .bind::<Uuid, _>(nat_entry.id)
+            .bind::<BigInt, _>(nat_entry.version_added)
+            .execute_async(&*self.pool_connection_authorized(opctx).await?)
+            .await
+            .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))?;
+
+        if updated_rows == 0 {
+            return Err(Error::ObjectNotFound {
+                type_name: ResourceType::Ipv4NatEntry,
+                lookup_type: LookupType::ByCompositeId(
+                    "id, version_added".to_string(),
+                ),
+            });
+        }
+        Ok(())
+    }
+
+    pub async fn ipv4_nat_cancel_delete(
```

Comment:
Could this be marked `#[cfg(test)]`?  The act of cancelling a delete seems somewhat unsafe to me outside test contexts.

Reply:
This was originally intended for usage in saga unwinds, but I think we've since decided that we shouldn't try to unwind deletion of network configuration and instead consider it the "point of no return"